docs/content/guides/dagster-pipes/subprocess/modify-external-code.mdx

```diff
@@ -0,0 +1,248 @@
+import pandas as pd
+from dagster_pipes import PipesContext, open_dagster_pipes
+
+
+def main():
+    orders_df = pd.DataFrame({"order_id": [1, 2], "item_id": [432, 878]})
+    total_orders = len(orders_df)
+    # get the Dagster Pipes context
+    context = PipesContext.get()
+    context.info.log(f"processing total {total_orders} orders")
+
+
+if __name__ == "__main__":
+    # connect to Dagster Pipes
+    with open_dagster_pipes():
+        main()
+```
+
+Then, the log messages will show up in the [**Run details**](/concepts/webserver/ui#run-details) page of the Dagster UI. You can filter the log levels to only view `info` level messages:
+
+1. Click the **Levels** filter next to the log filter field. This will present a dropdown of all log levels.
+2. Select the **info** checkbox and deselect the others. This will show only the logs marked as **info** level.
+
+<Image
+alt="Send log messages to Dagster"
+src="/images/guides/dagster-pipes/modifying-code-4-log-messages.png"
+width={1326}
+height={742}
+/>
+
+---
+
+## Step 3: Send structured metadata to Dagster
+
+Sometimes, you may want to log information from your external code as structured metadata shown in the Dagster UI. Dagster Pipes context also comes with the ability to log structured metadata back to Dagster.
+
+#### Report asset materialization
+
+Similar to [reporting materialization metadata within the Dagster process](/concepts/assets/software-defined-assets#recording-materialization-metadata), you can also report asset materialization back to Dagster from the external process.
+
+In this example, we’re passing a piece of metadata named `total_orders` to the `metadata` parameter of the <PyObject module="dagster_pipes" object="PipesContext" method="report_asset_materialization" />. This payload will be sent from the external process back to Dagster:
+
+```python file=/guides/dagster/dagster_pipes/subprocess/external_code_with_events.py startafter=start_events_marker endbefore=end_events_marker dedent=4
+# get the Dagster Pipes context
+context = PipesContext.get()
+# send structured metadata back to Dagster
+context.report_asset_materialization(metadata={"total_orders": total_orders})
+```
+
+Then, `total_orders` will show up in the UI as structured metadata:
+
+<Image
+alt="Report asset materialization to Dagster"
+src="/images/guides/dagster-pipes/modifying-code-5-1-report-asset-materialization.png"
+width={1412}
+height={806}
+/>
+
+This metadata will also be displayed on the **Events** tab of the **Asset Details** page in the UI:
+
+<Image
+alt="View materialization events in asset details page"
+src="/images/guides/dagster-pipes/modifying-code-5-2-materialization-asset-details.png"
+width={1286}
+height={736}
+/>
+
+#### Report asset checks
+
+Dagster allows you to define and execute data quality checks on assets. Refer to the [Asset Checks](/concepts/assets/asset-checks) documentation for more information.
```

Comment:
The `@asset` needs to have an `AssetCheckSpec` attached to it in order to report asset checks. This is referenced with "If your asset has data quality checks defined," but doesn't give much direction as to how to add them. I imagine that's because adding the asset check is something that would be done by whoever is responsible for part 1 of the tutorial, and not the stakeholder persona for this part. 

Should the AssetCheckSpec get added into the "create subprocess asset" tutorial part so that the code snippet added here works?

I'm also struggling to figure out exactly how to add the asset check spec to the `@asset` just doing this, like in the completed snippet at the bottom of the page
```python
@asset(
    check_specs=[AssetCheckSpec("no_empty_order_check", asset="subprocess_asset")]
def subprocess_asset():
    ....
)
```
results in 
```
DagsterInvariantViolationError: Expected Tuple annotation for multiple outputs, but received non-tuple annotation.
```

Reply:
updated the format to include dagster_code but this also seems to be a separate issue with the return type - will follow up on a later PR!